tap_redshift/__init__.py

```diff
@@ -0,0 +1,421 @@
+import datetime
+import copy
+import time
+import pendulum
+
+import psycopg2
+from itertools import groupby
+
+import singer
+import singer.metrics as metrics
+from singer import utils
+from singer.schema import Schema
+from singer.catalog import Catalog, CatalogEntry
+
+from . import resolve
+
+
+__version__ = '1.0.0'
+
+LOGGER = singer.get_logger()
+
+REQUIRED_CONFIG_KEYS = [
+    'host',
+    'port',
+    'dbname',
+    'user',
+    'password'
+]
+
+STRING_TYPES = set([
+    'char',
+    'enum',
+    'longtext',
+    'mediumtext',
+    'text',
+    'varchar'
+])
+
+BYTES_FOR_INTEGER_TYPE = {
+    'tinyint': 1,
+    'int2': 2,
+    'mediumint': 3,
+    'int4': 4,
+    'int8': 8
+}
+
+FLOAT_TYPES = set(['float', 'double'])
+
+DATETIME_TYPES = set(['datetime', 'timestamptz', 'date', 'time'])
+
+
+def discover_catalog(**kwargs):
+    '''Returns a Catalog describing the structure of the database.'''
+
+    # For testing purpose
+    if kwargs:
+        args = kwargs.get('mock')
+        dbname = args['dbname']
+    else:
+        args = utils.parse_args(REQUIRED_CONFIG_KEYS)
+        dbname = args.config['dbname']
+
+    table_spec = select_all("""
+       SELECT table_type, table_name
+       FROM INFORMATION_SCHEMA.Tables
+       WHERE table_schema = 'public'
+    """, **kwargs)
+
+    column_specs = select_all("""
+        SELECT c.table_name, c.ordinal_position, c.column_name, c.udt_name
+        FROM INFORMATION_SCHEMA.Tables t
+        JOIN INFORMATION_SCHEMA.Columns c ON c.table_name = t.table_name
+        WHERE t.table_schema = 'public'
+        ORDER BY c.table_name, c.ordinal_position
+    """, **kwargs)
+
+    entries = []
+    column = [{'name': k, 'columns': [
+                {'pos': t[1], 'name': t[2], 'type': t[3]} for t in v]}
+              for k, v in groupby(column_specs, key=lambda t: t[0])]
+
+    for items in column:
+        table_name = items['name']
+        cols = items['columns']
+        schema = Schema(type='object',
+                        properties={
+                            c['name']: schema_for_column(c) for c in cols})
+        tap_stream_id = '{}-{}'.format(dbname, table_name)
+        entry = CatalogEntry(
```

Comment:
I've added both `key_properties` and `metadata` properties. As for the `replication_key` and  `replication_method`, I think it's tricky to have them added as part of the `CatalogEntry` property when in discovery mode. From my level of understanding from the best practises and the tap-mysql, those properties are optional and the user can easily add them to the properties.json when in sync mode since we can't guess which `replication_method` the user will want to use. That way, the user can specify what `replication_method` (FULL or INCREMENTAL) and also keys that are unique to the schema that can be used as the `replication_key`.